[FSDP] Clean up `_lazy_init()`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81243 [FSDP] Move `_post_backward_called` to `_init_param_attributes`
* **#80185 [FSDP] Clean up `_lazy_init()`**
* #80087 [Easy][FSDP] Add `zero_grad()` to unit test train loop
* #80002 [FSDP] Remove `self.numel_padded_per_param` (unused)
* #80000 [FSDP] Move tensor sharding logic to `FlatParamHandle`
* #79984 [FSDP] Deduplicate `_orig_size` and `_unsharded_size`
* #79652 [FSDP] Introduce `FlatParamHandle`

This PR cleans up `_lazy_init()`. The explanations are left as PR comments.

Differential Revision: [D37726059](https://our.internmc.facebook.com/intern/diff/D37726059)

Commit:
[FSDP] Clean up `_lazy_init()`

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1461,7 +1461,6 @@ def __getitem__(self, key: int) -> Any:
     def _reset_lazy_init(self) -> None:
         """
         Reset instance so :func:`_lazy_init` will run on the next forward.
-        Currently this is only called in __init__
```
Removing this comment since it is not true. It is also called in `apply()`.

nit: I think as a general rule, we should avoid these kinds of comments since they are not maintainable. Someone who adds a call to `_reset_lazy_init()` somewhere may not know to update this comment.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2287,6 +2250,7 @@ def _load_sharded_state_dict(
     def forward(self, *args: Any, **kwargs: Any) -> Any:
         with torch.autograd.profiler.record_function("FullyShardedDataParallel.forward"):
             self._lazy_init()
+            self._wait_for_previous_optim_step()
```
I separated out `_wait_for_previous_optim_step()`, which performs *synchronization*, from `_lazy_init()` since it does not logically match.

`_lazy_init()` has been refactored to only run its actual logic once, meaning that it is a no-op after the first `forward()`. Before, the implementation still relied on the `_wait_for_previous_optim_step()` synchronization even though the actual initialization logic in `_lazy_init()` was a no-op.

Other methods like `state_dict()` and `summon_full_params()` call `_lazy_init()` but require hard synchronization using `torch.cuda.synchronize()`, so those do not need the additional `_wait_for_previous_optim_step()` call, further motivating the separation.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1479,38 +1478,46 @@ def _reset_lazy_init(self) -> None:
+        This method's actual logic only runs on the root FSDP instance, which
+        performs initialization for all non-root FSDP instances to avoid
+        partial initialization.
+        """
```
There were two components to the lazy initialization:
1. Setting each FSDP instance's `_is_root` and sharing some data structures (streams, execution order info, graph order) from the root instance to non-root instances.
2. Initializing parameter attributes (via `_init_param_attributes()`).

For 1, the root FSDP instance does it all at once, meaning it is done when the root FSDP instance's `_lazy_init()` returns. In contrast, for 2, each FSDP instance is responsible for itself.

This exposes a window between the end of the root's `_lazy_init()` and a non-root's `_lazy_init()`, where the non-root is partially initialized. There is no performance downside to this, but logically, I think it is clearer if the initialization is done atomically with respect to one `_lazy_init()` call.

This means that `self._is_root is None` fully indicates whether an FSDP instance's initialization has happened or not. Before, for a non-root FSDP instance, it could be that `self._is_root == False`, but `_init_param_attributes()` had not been called yet.

This also means that we have a clear separation between the logic that is expected to run every iteration (e.g. `_wait_for_previous_optim_step()`) versus the one-time initialization logic.

awgu replied:
The early return in `_init_param_attributes()` can actually be removed now since it is only ever called once, but since there is an `assert` for CPU offloading, I decided to leave it in.

awgu reviewed (commented):
Explanations in the comments.

rohan-varma reviewed (approved):
I think this might have a small semantic change, where cast_buffers() is no longer done every iteration - I think this is fine since it was probably redundant. But one case I can think of is we checkpoint in full precision for buffers. If user calls forward after that, we should make sure they were restored (I think they are already when exiting checkpoint logic, but need to check)

awgu commented:
> I think this might have a small semantic change, where cast_buffers() is no longer done every iteration - I think this is fine since it was probably redundant. But one case I can think of is we checkpoint in full precision for buffers. If user calls forward after that, we should make sure they were restored (I think they are already when exiting checkpoint logic, but need to check)

It looks like there is no cast in the post load state dict hooks. I will test this and add the cast to the post load state dict hook as needed. IMO, it makes more sense (for both clarity and performance overhead) to have that cast once and directly where we need it rather than having it be a no-op all but one of the time when calling it in `_lazy_init()` every iteration.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1479,38 +1478,46 @@ def _reset_lazy_init(self) -> None:
+        # Do not reshard the root's parameters at the end of the forward pass
+        # with the intention that they are immediately used in the backward
+        # pass gradient computation (though this may not be true)
+        self.reshard_after_forward = False
```
I do not think so.

The places where `self.reshard_after_forward` is set are as follows:
- Old `_lazy_init()`:
https://github.com/pytorch/pytorch/blob/10c689856d5f913b93daa404955927a5a5716805/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1492-L1503
- `_init_reshard_after_forward()`:
https://github.com/pytorch/pytorch/blob/10c689856d5f913b93daa404955927a5a5716805/torch/distributed/fsdp/fully_sharded_data_parallel.py#L958-L973
- `_init_reshard_after_forward()` is only called in `_reset_lazy_init()`:
https://github.com/pytorch/pytorch/blob/10c689856d5f913b93daa404955927a5a5716805/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1457-L1473

Thus, `self.reshard_after_forward` is only set in `_reset_lazy_init()` (based on sharding strategy) and set to `False` for the root in `_lazy_init()`.

This PR does not touch `_reset_lazy_init()` and preserves the fact that the root's `self.reshard_after_forward` is set to `False` in `_lazy_init()`.

awgu commented:
> I think this might have a small semantic change, where cast_buffers() is no longer done every iteration - I think this is fine since it was probably redundant. But one case I can think of is we checkpoint in full precision for buffers. If user calls forward after that, we should make sure they were restored (I think they are already when exiting checkpoint logic, but need to check)

https://github.com/pytorch/pytorch/blob/10c689856d5f913b93daa404955927a5a5716805/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1909-L1930

The `_post_state_dict_hook()` casts buffers.

https://github.com/pytorch/pytorch/blob/10c689856d5f913b93daa404955927a5a5716805/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1992-L2005

`state_dict()` is the only method that casts buffers (namely, back to the original dtype) other than `_cast_buffers()`.

Since `state_dict()` and `_post_state_dict_hook()` always are called in pairs, there should be no need to `_cast_buffers()` every iteration.

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1479,38 +1478,46 @@ def _reset_lazy_init(self) -> None:
+        self._is_root = True
+        self._assert_state(TrainingState_.IDLE)
+        self._init_streams()
+        self._cast_buffers(recurse=True)
```
https://github.com/pytorch/pytorch/pull/80185#issuecomment-1179305309

Commit:
Update on "[FSDP] Clean up `_lazy_init()`"


This PR cleans up `_lazy_init()`. The explanations are left as PR comments.

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726059).

Commit:
Update on "[FSDP] Clean up `_lazy_init()`"


This PR cleans up `_lazy_init()`. The explanations are left as PR comments.

Differential Revision: [D37726059](https://our.internmc.facebook.com/intern/diff/D37726059)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726059).

Commit:
Update on "[FSDP] Clean up `_lazy_init()`"


This PR cleans up `_lazy_init()`. The explanations are left as PR comments.

Differential Revision: [D37726059](https://our.internmc.facebook.com/intern/diff/D37726059)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726059).

awgu commented:
@pytorchbot merge
